### PR TITLE
SLVS-1131 Local instance method calls should reset constraints on loc…

### PR DESCRIPTION
**Pull request by valhristov**
…al fields

### CODE REVIEW

File: src/SonarAnalyzer.Common/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 1]
```diff
@@ -293,6 +291,16 @@ internal bool IsSymbolTracked(ISymbol symbol)
                 symbol.ContainingSymbol.Equals(declaration);
         }
 
+        protected bool IsFieldSymbol(ISymbol symbol)
+        {
+            var field = symbol as IFieldSymbol;
+
+            return field != null &&
+                declaration.ContainingType
+                    .GetSelfAndBaseTypes()
```
Comment (Evangelink) [inline src/SonarAnalyzer.Common/Helpers/SymbolicExecution/ExplodedGraph.cs]: That's personal taste but I prefer to have aligned chained call. You are free to ignore this comment.

File: src/SonarAnalyzer.Common/Helpers/SymbolicExecution/ProgramState.cs [hunk 2]
```diff
@@ -316,6 +316,11 @@ public SymbolicValue GetSymbolValue(ISymbol symbol)
             return null;
         }
 
+        public ProgramState RemoveSymbolValues(IEnumerable<ISymbol> symbols)
```
Comment (Evangelink) [inline src/SonarAnalyzer.Common/Helpers/SymbolicExecution/ProgramState.cs]: I would remove this method and do the inner call directly from the calling method. I think this middle method makes it harder to understand.

File: src/SonarAnalyzer.Common/Helpers/SymbolicExecution/ProgramState.cs [hunk 3]
```diff
@@ -340,9 +345,14 @@ internal int GetVisitedCount(ProgramPoint programPoint)
 
         internal ProgramState Clean(IEnumerable<ISymbol> liveSymbolsToKeep)
         {
+            return Clean(symbol => liveSymbolsToKeep.Contains(symbol) || symbol is IFieldSymbol);
+        }
+
+        internal ProgramState Clean(Func<ISymbol, bool> shouldKeepSymbol)
+        {
             // TODO: Remove the OR part when SLVS-1136 is fixed
```
Comment (Evangelink) [inline src/SonarAnalyzer.Common/Helpers/SymbolicExecution/ProgramState.cs]: Move this comment to line 348 as the IFieldSymbol is no longer done here.

File: src/SonarAnalyzer.Common/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 4]
```diff
@@ -260,7 +260,10 @@ protected ProgramState CleanStateAfterBlock(ProgramState programState, Block blo
         {
             var liveVariables = lva.GetLiveOut(block)
                 .Union(nonInDeclarationParameters); // LVA excludes out and ref parameters
-            return programState.Clean(liveVariables);
+
+            // TODO: Remove the IFieldSymbol check when SLVS-1136 is fixed
```
Comment (SonarTech) [inline src/SonarAnalyzer.Common/Helpers/SymbolicExecution/ExplodedGraph.cs]: INFO Complete the task associated to this "TODO" comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* INFO 1 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'

**Pull request merged**